File: ki-entities/websites/capital-one.lua
----------------------------------------------------------------------------------------------------
-- Capital One website config
--
local Ki = spoon.Ki
local Website = Ki.Website
local CapitalOne = Website { "CapitalOne", "https://www.capitalone.com" }

-- Retrieve minified offer scraping code from `~/.hammerspoon/scripts/offers`: `npm run capital-one`
function CapitalOne.getOffers()
    return hs.osascript.applescript([[
        set info to ""

        tell application "Safari" to tell current tab of front window
            set info to do JavaScript "window.rows=[];var getTarget=(e,r,t)=>{e=e.querySelector(r);return t?e?e[t]:null:e},offerItems=(Date.prototype.addDays=function(e){var r=new Date(this.valueOf());return r.setDate(r.getDate()+e),r},Array.from(document.querySelectorAll(`.item`))),parseOffer=e=>{var r,e=getTarget(e,`a[role='button'] > div`,`innerText`);return e?([,r=e]=e.match(/Get (.+) back/)||[],r.includes(`up to`)?([,e=e]=r.match(/up to (.+)/)||[],e):r):``};window.rows.push(...offerItems.map(e=>({card:`Capital One Savor One`,name:getTarget(e,`.merchantCleanName`,`innerText`),offer:parseOffer(e),expiry:getTarget(e,`:scope div > div > div`,`innerText`)})).map(e=>[e.card,e.name,e.offer,e.expiry,``].join(`\\t`))),window.rows.join(`\\n`);" as text
        end tell

        return info
    ]])
end

function CapitalOne:copyOffers()
    local isOk, output = self.getOffers()
    if isOk then
        hs.pasteboard.setContents(output)
        hs.notify.show("Ki", "Copied Capital One offers to clipboard", "")
    else
        hs.notify.show("Ki", "Unable to copy Capital One offers to clipboard", "")
    end
end

CapitalOne:registerShortcuts({
    { nil, "c", function() CapitalOne:copyOffers() end },
})

return CapitalOne

File: scripts/offers/capital-one.js
// Initialize row headers
window.rows = [];

// Helper functions
var getTarget = (container, selector, attribute) => {
    const target = container.querySelector(selector);

    if (attribute) {
        return target ? target[attribute] : null;
    }

    return target;
};
Date.prototype.addDays = function(days) {
    const date = new Date(this.valueOf());
    date.setDate(date.getDate() + days);
    return date;
}

var offerItems = Array.from(document.querySelectorAll('.item'));

var parseOffer = container => {
    const offerText = getTarget(container, `a[role='button'] > div`, 'innerText');
    if (!offerText) {
        return '';
    }

    let [, offer = offerText] = offerText.match(/Get (.+) back/) || [];
    if (offer.includes('up to')) {
        let [, upToOffer = offerText] = offer.match(/up to (.+)/) || [];
        return upToOffer;
    }

    return offer;
};

window.rows.push(
    ...offerItems
        .map((offerItem) => ({
            card: 'Capital One Savor One',
            name: getTarget(offerItem, '.merchantCleanName', 'innerText'),
            offer: parseOffer(offerItem),
            expiry: getTarget(offerItem, ':scope div > div > div', 'innerText'),
        }))
        .map(data => [
            data.card,
            data.name,
            data.offer,
            data.expiry,
            '',
        ].join('\t'))
);

window.rows.join('\n');
